pkgs/by-name/README.md

```diff
@@ -76,6 +76,17 @@ and override its value in [`pkgs/top-level/all-packages.nix`](../top-level/all-p
 }
 ```
 
+### Toplevel argument restrictions
+The list of toplevel arguments should be restricted to packages defined in the
+common nixpkgs package set, and simple `enable*` boolean flags.
+
+As the expression is being `callPackage`'ed, both newly-defined packages, as
+well as those added through the use of overlays will cause defaults to be
+overwritten in a unexpected and hard-to-debug fashion.
+
+In case you're intending to write more flexible package expressions, consider
```

Comment:
This is not `by-name`'s mistake. This is about `callPackage`. Whenever people use `callPackage` they are affected.

Reply:
Agreed it's not `by-name`'s mistake.

This also say it's because it's callPackage'd. I still think it's worth to write it down like this, because people tend to be too fancy with what they do in `by-name`, and restricting it only to packages defined in the top-level (without defaults) and to only boolean enable flags prevents this effectively.

Reply:
The main restriction with by-name is that you can't use a custom callPackage, there's nothing else that I know of affecting the override interface. Besides, there is an upcoming check to enforce moving to by-name when touching packages at the top-level, so this suggestion doesn't make sense.

Reply:
Please propose exact wording in a new PR.